doc/src/devdocs/inference.md

```diff
@@ -0,0 +1,98 @@
+
+# Inference
```

Comment:
Perhaps we should add links to https://juliacomputing.com/blog/2016/04/04/inference-convergence.html and https://juliacomputing.com/blog/2017/05/15/inference-converage2.html to describe the overall structure?

I should also probably add a description of https://github.com/JuliaLang/julia/pull/21933 here too, later, when I get a chance.